main/exercise/ptest_category.class.php

```diff
@@ -0,0 +1,1174 @@
+                    exercice_id = $exerciseId AND
+                    qrc.c_id = $courseId
+                ";
+
+        $res = Database::query($sql);
+        $categories = [];
+        while ($data = Database::fetch_array($res)) {
+            if (!empty($check_in_question_list)) {
+                if (!in_array($data['question_id'], $check_in_question_list)) {
+                    continue;
+                }
+            }
+
+            if (!isset($categories[$data['category_id']]) ||
+                !is_array($categories[$data['category_id']])
+            ) {
+                $categories[$data['category_id']] = [];
+            }
+
+            $categories[$data['category_id']][] = $data['question_id'];
+        }
+
+        if (!empty($categoriesAddedInExercise)) {
+            $newCategoryList = [];
+            foreach ($categoriesAddedInExercise as $category) {
+                $categoryId = $category['category_id'];
+                if (isset($categories[$categoryId])) {
+                    $newCategoryList[$categoryId] = $categories[$categoryId];
+                }
+            }
+
+            $checkQuestionsWithNoCategory = false;
+            foreach ($categoriesAddedInExercise as $category) {
+                if (empty($category['category_id'])) {
+                    // Check
+                    $checkQuestionsWithNoCategory = true;
+                    break;
+                }
+            }
+
+            // Select questions that don't have any category related
+            if ($checkQuestionsWithNoCategory) {
+                $originalQuestionList = $check_in_question_list;
+                foreach ($originalQuestionList as $questionId) {
+                    $categoriesFlatten = array_flatten($categories);
+                    if (!in_array($questionId, $categoriesFlatten)) {
+                        $newCategoryList[0][] = $questionId;
+                    }
+                }
+            }
+            $categories = $newCategoryList;
+        }
+
+        return $categories;
+    }
+
+    /**
+     * Returns an array of $numberElements from $array.
+     *
+     * @param array
+     * @param int
+     *
+     * @return array
+     */
+    public static function getNElementsFromArray($array, $numberElements)
+    {
+        $list = $array;
+        shuffle($list);
+        if ($numberElements < count($list)) {
+            $list = array_slice($list, 0, $numberElements);
+        }
+
+        return $list;
+    }
+
+    /**
+     * @param int $questionId
+     * @param int $displayCategoryName
+     */
+    public static function displayCategoryAndTitle($questionId, $displayCategoryName = 1)
+    {
+        echo self::returnCategoryAndTitle($questionId, $displayCategoryName);
+    }
+
+    /**
+     * @param int $questionId
+     * @param int $in_display_category_name
+     *
+     * @return string|null
+     */
+    public static function returnCategoryAndTitle($questionId, $in_display_category_name = 1)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The variable $is_student is not named in camelCase.](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5018478)